Jaro-Winkler-Distance considers the length of strings, the number of matching characters and the number of transpositions, also we use a prefix scale that gives us more accurate answers when strings have a common prefix 

In [1]:
import logging

import jellyfish
import pandas as pd
import numpy as np

logger = logging.getLogger(__name__)

In [2]:
df_dbpedia = pd.read_csv("./files/dbpedia_classes.csv")
df_imageNet = pd.read_csv("./files/ImageNet.csv")
df_imageNet['Class Name'] = df_imageNet['Class Name'].str.replace(", ",",")
df_imageNet['Class Name'] = df_imageNet['Class Name'].str.split(",")

In [3]:
def get_jaro_winkler_distance():
    max_ratio_list = []
    max_label_list = []
    for class_list in df_imageNet['Class Name']:
        max_val = 0
        max_label = ''
        for item in class_list:
            for label in df_dbpedia['label']:
                if max_val< jellyfish.jaro_winkler_similarity(item, label):
                    max_val = jellyfish.jaro_winkler_similarity(item, label)
                    max_label = label
        max_ratio_list.append(max_val)
        max_label_list.append(max_label)
        
    df_imageNet['jaro_winkler_distance'] = max_ratio_list
    df_imageNet['dbpedia_class'] = max_label_list
    df_imageNet.rename(columns = {"Class Name":"imageNet_class"}, inplace=True)
            
get_jaro_winkler_distance()

In [4]:
df_imageNet


Class ID                                     imageNet_class  \
0           0                               [tench, Tinca tinca]   
1           1                      [goldfish, Carassius auratus]   
2           2  [great white shark, white shark, man-eater, ma...   
3           3                  [tiger shark, Galeocerdo cuvieri]   
4           4                     [hammerhead, hammerhead shark]   
..        ...                                                ...   
995       995                                        [earthstar]   
996       996  [hen-of-the-woods, hen of the woods, Polyporus...   
997       997                                           [bolete]   
998       998                            [ear, spike, capitulum]   
999       999     [toilet tissue, toilet paper, bathroom tissue]   

     jaro_winkler_distance  dbpedia_class  
0                 0.790000         tenure  
1                 0.693182    golf course  
2                 0.844444            man  
3                 0.762756  figure skater  
4                 0.777778         camera  
..                     ...            ...  
995               0.853333     earthquake  
996               0.715278         island  
997               0.800000           bone  
998               0.916667           year  
999               0.766026  tennis player  

[1000 rows x 4 columns]

Lets look at the number of classes with a distance score equal to 100

In [5]:
result =  df_imageNet[df_imageNet['jaro_winkler_distance' ]== 1.0]
result


Class ID                                     imageNet_class  \
134       134                                      [crane, bird]   
242       242                                            [boxer]   
286       286  [cougar, puma, catamount, mountain lion, paint...   
323       323  [monarch, monarch butterfly, milkweed butterfl...   
408       408                    [amphibian, amphibious vehicle]   
437       437         [beacon, lighthouse, beacon light, pharos]   
483       483                                           [castle]   
487       487  [cellular telephone, cellular phone, cellphone...   
497       497                          [church, church building]   
525       525                                  [dam, dike, dyke]   
553       553               [file, file cabinet, filing cabinet]   
624       624                                          [library]   
663       663                                        [monastery]   
668       668                                           [mosque]   
687       687                                [organ, pipe organ]   
705       705                   [passenger car, coach, carriage]   
743       743                             [prison, prison house]   
762       762   [restaurant, eating house, eating place, eatery]   
812       812                                    [space shuttle]   
829       829   [streetcar, tram, tramcar, trolley, trolley car]   
916       916           [web site, website, internet site, site]   
979       979                                     [valley, vale]   
980       980                                          [volcano]   
981       981                      [ballplayer, baseball player]   

     jaro_winkler_distance    dbpedia_class  
134                    1.0             bird  
242                    1.0            boxer  
286                    1.0          painter  
323                    1.0          monarch  
408                    1.0        amphibian  
437                    1.0       lighthouse  
483                    1.0           castle  
487                    1.0     mobile phone  
497                    1.0           church  
525                    1.0              dam  
553                    1.0             file  
624                    1.0          library  
663                    1.0        monastery  
668                    1.0           mosque  
687                    1.0            organ  
705                    1.0            coach  
743                    1.0           prison  
762                    1.0       restaurant  
812                    1.0    space shuttle  
829                    1.0        streetcar  
916                    1.0          website  
979                    1.0           valley  
980                    1.0          volcano  
981                    1.0  baseball player

In [6]:
print(result.shape)

(24, 4)


Lets look at the number of classes with a distance score higher than 0.9

In [7]:
result =  df_imageNet[df_imageNet['jaro_winkler_distance' ] > 0.9]
result

Class ID                                     imageNet_class  \
22         22  [bald eagle, American eagle, Haliaeetus leucoc...   
64         64                                      [green mamba]   
132       132  [American egret, great white heron, Egretta al...   
134       134                                      [crane, bird]   
231       231                                           [collie]   
..        ...                                                ...   
979       979                                     [valley, vale]   
980       980                                          [volcano]   
981       981                      [ballplayer, baseball player]   
989       989                           [hip, rose hip, rosehip]   
998       998                            [ear, spike, capitulum]   

     jaro_winkler_distance    dbpedia_class  
22                0.931429  American Leader  
64                0.905455       green alga  
132               0.914762  American Leader  
134               1.000000             bird  
231               0.909524          college  
..                     ...              ...  
979               1.000000           valley  
980               1.000000          volcano  
981               1.000000  baseball player  
989               0.916667             ship  
998               0.916667             year  

[103 rows x 4 columns]

Lets look at the number of classes with a distance score higher than 0.8

In [8]:
result =  df_imageNet[df_imageNet['jaro_winkler_distance' ] > 0.8]
result

Class ID                                     imageNet_class  \
2           2  [great white shark, white shark, man-eater, ma...   
5           5       [electric ray, crampfish, numbfish, torpedo]   
7           7                                             [cock]   
10         10              [brambling, Fringilla montifringilla]   
11         11                   [goldfinch, Carduelis carduelis]   
..        ...                                                ...   
988       988                                            [acorn]   
989       989                           [hip, rose hip, rosehip]   
990       990                  [buckeye, horse chestnut, conker]   
995       995                                        [earthstar]   
998       998                            [ear, spike, capitulum]   

     jaro_winkler_distance          dbpedia_class  
2                 0.844444                    man  
5                 0.841905  electrical substation  
7                 0.833333                   lock  
10                0.896296                  brain  
11                0.835133     Cardinal direction  
..                     ...                    ...  
988               0.893333                  actor  
989               0.916667                   ship  
990               0.894444                conifer  
995               0.853333             earthquake  
998               0.916667                   year  

[486 rows x 4 columns]

Lets look at the number of classes with a distance score higher than 0.7

In [9]:
result =  df_imageNet[df_imageNet['jaro_winkler_distance' ] > 0.7]
result

Class ID                                     imageNet_class  \
0           0                               [tench, Tinca tinca]   
2           2  [great white shark, white shark, man-eater, ma...   
3           3                  [tiger shark, Galeocerdo cuvieri]   
4           4                     [hammerhead, hammerhead shark]   
5           5       [electric ray, crampfish, numbfish, torpedo]   
..        ...                                                ...   
995       995                                        [earthstar]   
996       996  [hen-of-the-woods, hen of the woods, Polyporus...   
997       997                                           [bolete]   
998       998                            [ear, spike, capitulum]   
999       999     [toilet tissue, toilet paper, bathroom tissue]   

     jaro_winkler_distance          dbpedia_class  
0                 0.790000                 tenure  
2                 0.844444                    man  
3                 0.762756          figure skater  
4                 0.777778                 camera  
5                 0.841905  electrical substation  
..                     ...                    ...  
995               0.853333             earthquake  
996               0.715278                 island  
997               0.800000                   bone  
998               0.916667                   year  
999               0.766026          tennis player  

[910 rows x 4 columns]

Conclusion: We could use jaro_winkler distance but we need to set the treshold very high because with > 80 we already have 910 out of all classes.
A problem here is that it only compares strings not semantics (for example Class ID: 999 has a 0,76 distance but is actually not the same),
also there is no significant change between choosing jaro and jaro_winkler



Idea: Right now we take the highest value out of the image_net classes array, but lets change it so we take the average of the values as distance for each class and assign the class with the highest value to the dbpedia class